// @osdoc-dev/avenger-utils/lib/generator.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fileGenerator = void 0;
var path_1 = __importDefault(require("path"));
var file_1 = require("./file");
var index_1 = require("./index");
var copyTpl = function (opts) { return __awaiter(void 0, void 0, void 0, function () {
    var tpl, content;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                tpl = index_1.fs.readFileSync(opts.templatePath, 'utf-8');
                content = index_1.Mustache.render(tpl, opts.context);
                return [4, file_1.createFolder(path_1.default.dirname(opts.target))];
            case 1:
                _a.sent();
                index_1.log(opts.target + " \u5199\u5165\u6210\u529F\uFF01");
                index_1.fs.writeFileSync(opts.target, content, 'utf-8');
                return [2];
        }
    });
}); };
var fileGenerator = function (opt) {
    var _a = opt || {}, target = _a.target, source = _a.source, context = _a.context, _b = _a.ignores, ignores = _b === void 0 ? [] : _b;
    var files = index_1.glob.sync('**/*', {
        cwd: source,
        dot: true,
        ignore: ['**/node_modules/**'],
    });
    files.forEach(function (file) {
        if (ignores.includes(file))
            return;
        var absFile = path_1.default.join(source, file);
        if (index_1.fs.statSync(absFile).isDirectory())
            return;
        if (file.endsWith('.tpl')) {
            copyTpl({
                templatePath: absFile,
                target: path_1.default.join(target, file.replace(/\.tpl$/, '')),
                context: context,
            });
        }
        else {
            var absTarget = path_1.default.join(target, file);
            file_1.createFolder(path_1.default.dirname(absTarget));
            index_1.log(absTarget + " \u5199\u5165\u6210\u529F\uFF01");
            index_1.fs.copyFileSync(absFile, absTarget);
        }
    });
};
exports.fileGenerator = fileGenerator;
