// @osdoc-dev/avenger-utils/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fs = exports.glob = exports.Mustache = exports.prettier = exports.inquirer = exports.shelljs = exports.rimraf = exports.lodash = exports.slash = exports.chalk = exports.semver = void 0;
var semver_1 = __importDefault(require("semver"));
exports.semver = semver_1.default;
var chalk_1 = __importDefault(require("chalk"));
exports.chalk = chalk_1.default;
var slash2_1 = __importDefault(require("slash2"));
exports.slash = slash2_1.default;
var lodash_1 = __importDefault(require("lodash"));
exports.lodash = lodash_1.default;
var rimraf_1 = __importDefault(require("rimraf"));
exports.rimraf = rimraf_1.default;
var shelljs_1 = __importDefault(require("shelljs"));
exports.shelljs = shelljs_1.default;
var inquirer_1 = __importDefault(require("inquirer"));
exports.inquirer = inquirer_1.default;
var prettier_1 = __importDefault(require("prettier"));
exports.prettier = prettier_1.default;
var mustache_1 = __importDefault(require("mustache"));
exports.Mustache = mustache_1.default;
var glob_1 = __importDefault(require("glob"));
exports.glob = glob_1.default;
var fs_extra_1 = __importDefault(require("fs-extra"));
exports.fs = fs_extra_1.default;
__exportStar(require("./validate"), exports);
__exportStar(require("./log"), exports);
__exportStar(require("./package"), exports);
__exportStar(require("./file"), exports);
__exportStar(require("./generator"), exports);
