// @osdoc-dev/avenger-cli/lib/common.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCreateArguments = exports.getBuildArguments = exports.checkNodeVersion = void 0;
var avenger_utils_1 = require("@osdoc-dev/avenger-utils");
var minimist_1 = __importDefault(require("minimist"));
var avenger_shared_1 = require("@osdoc-dev/avenger-shared");
var checkNodeVersion = function (needVerison, packageName) {
    if (!avenger_utils_1.semver.satisfies(process.version, needVerison, { includePrerelease: true })) {
        avenger_utils_1.error("You are using Node " + process.version + ", but " + packageName + " " +
            ("requires Node " + needVerison + ".\nPlease upgrade your Node version."));
        process.exit(1);
    }
};
exports.checkNodeVersion = checkNodeVersion;
var getBuildArguments = function () {
    var opt = minimist_1.default(process.argv.slice(3));
    var _a = opt, file = _a.file, entry = _a.entry, watch = _a.watch, esm = _a.esm, cjs = _a.cjs, umd = _a.umd;
    return {
        cwd: process.cwd(),
        watch: watch || false,
        inlineConfig: {
            outFile: file,
            entry: entry,
            esm: esm ? 'rollup' : false,
            cjs: cjs ? 'rollup' : false,
            umd: umd ? 'rollup' : false,
        },
    };
};
exports.getBuildArguments = getBuildArguments;
var getTemplate = function (template) {
    if (!template)
        return '';
    var temps = Object.keys(avenger_shared_1.CreateProjectType).map(function (v) { return v; });
    var ret = template;
    if (!temps.includes(template)) {
        avenger_utils_1.log("\u5F53\u524D template \u4E0D\u5728\u9884\u8BBE\u8303\u56F4\u5185\uFF0C\u5C06\u9ED8\u8BA4\u4F7F\u7528 " + avenger_shared_1.CreateProjectType.basic);
        ret = avenger_shared_1.CreateProjectType.basic;
    }
    return ret;
};
var getCreateArguments = function () {
    var opt = minimist_1.default(process.argv.slice(3));
    var _a = opt, template = _a.template, force = _a.force, _ = _a._, gitUrl = _a.git;
    return {
        name: _[0] || '',
        options: {
            force: force || false,
            template: getTemplate(template),
            git: gitUrl,
        },
    };
};
exports.getCreateArguments = getCreateArguments;
