// @osdoc-dev/avenger-config/lib/config-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var avenger_utils_1 = require("@osdoc-dev/avenger-utils");
var avenger_shared_1 = require("@osdoc-dev/avenger-shared");
var schema = avenger_utils_1.createSchema(function (joi) {
    return joi.object({
        esm: [
            joi.boolean(),
            joi.string(),
            joi.object({
                type: joi.string().valid(avenger_shared_1.BundleTypeMap.babel, avenger_shared_1.BundleTypeMap.rollup).required(),
                outFile: joi.string(),
                minify: joi.boolean(),
                sourcemap: joi.boolean(),
            }),
        ],
        cjs: [
            joi.boolean(),
            joi.string(),
            joi.object({
                type: joi.string().valid(avenger_shared_1.BundleTypeMap.babel, avenger_shared_1.BundleTypeMap.rollup).required(),
                outFile: joi.string(),
                minify: joi.boolean(),
                sourcemap: joi.boolean(),
            }),
        ],
        umd: [
            joi.boolean(),
            joi.object({
                outFile: joi.string(),
                minify: joi.boolean(),
                sourcemap: joi.boolean(),
                minFile: joi.boolean(),
                name: joi.string(),
                globals: joi.object(),
            }),
        ],
        outFile: joi.string(),
        entry: [joi.string()],
        disableTypeCheck: joi.boolean(),
        typescriptOpts: joi.object(),
        include: joi.string(),
        cssModule: [joi.boolean(), joi.object()],
        extraReplacePluginOpts: joi.object(),
        extraInjectPluginOpts: joi.object(),
        extraPostCssPluginOpt: joi.object(),
        extraPostCssOpt: joi.object(),
        autoprefixerOpts: joi.object(),
        extractCSS: joi.boolean(),
        injectCSS: joi.boolean(),
        rollupSassOpt: joi.object(),
        rollupLessOpt: joi.object(),
        runtimeHelpers: joi.boolean(),
        extraBabelPlugins: joi.object(),
        extraBabelPresets: joi.object(),
        extraExternal: joi.array().items(joi.string()),
    });
});
exports.default = schema;
